### ISSUE TITLE
Validation prior to saving

### ISSUE THREAD

**Issue by AlchemyUnited** [CMB2/CMB2#100]
Hello CMB2

This is my first time diving into to CMB2 so please type slower when you reply. Thanks :)

In short, I want validation (e.g., required, value > X, etc)  _prior_ to an DB update. By prior I mean I want to prevent publishing if there are validation issues. 

I'm combing through the code the best I can but maybe I missed something? The closest I can seem to find is in CMB2_field > function update_data(..)

This line: $override = apply_filters( 'cmb2_override_meta_save', null, $a, $this->args(), $this );

<sidebar>

I actually have my own (semi-experimental and) far _less_ robust CMB + form control generator. It doesn't have as many field types but it does have server side validation, support HTML5 validation and displays error messages, etc. If you can give me some insights - and perhaps a couple hooks, if necessary - I can consider trying to post it over to CMB2. Perhaps as a kinda add-on / step-brother plugin?

</sidebar>

Along the same lines, is there a reason the HTML5 validation stuff isn't supported? Perhaps there a js plugin you recommend for client side that is known to play nice with CMB2? And a tutorial perhaps?

Finally, all that said, the most effective way I've found for rendering form controls input is to treat them like strings. That is, since the attributes are value pairs you can just use '<' . implode() . '>' to create the string and you're done. Of course this example is somewhat over-simplified but you get the point. The advantage is, you can easily add data attributes and anything else that's typical to form controls. 

Just a couple extra cents that I thought might help.

p.s. The name key is actually the label. and the id becomes the name. I know you probably know this. But it's kinda confusing, yes?

**Comment by praditha** [comment]
Hi @jtsternberg ,
Thanks for point me out to the sanitization_cb, I just used it but if the value is invalid (regarding the validation rule) then the value on the form will be wiped out.
But how can I show the error message to the user rather then put a small information under the input field.

For example:
I have a reference code field which should be unique for every post.
If user put the code which already used by another post, I need to alert user that the code is already used.

Thanks.

**Comment by AlchemyUnited** [comment]
@praditha - Mind if I jump in? Off the top of my head (so forgive me if I'm mistaken)...

When you do the sanitization_cb, write the current value, as well as the new value to another meta row (e.g., my_cmb2_validation_errors). When displaying check that my_cmb2_validation_errors to see if you need to do anything error-oriented. For example, use the original value in my_cmb2_validation_errors not what's coming from that DB. my_cmb2_validation_errors can also be used to decide if you need to display an error message. 

Something like that? ;)

**Comment by jtsternberg** [comment]
Yep, that's a decent solution, then you could use the `cmb2_before_field_row` or `cmb2_after_field_row` hooks to output any found errors.

**Comment by AlchemyUnited** [comment]
@jtsternberg Decent? I guess I'll take that as a compliment ;)
